File: frontend/src/services/negotiationService.ts
import api from './api';
import { Offer } from '../types';

export interface CreateOfferPayload {
  productId: number;
  productName: string;
  buyerId: number;
  supplierId: number;
  proposedPrice: number;
  originalPrice: number;
  quantity: number;
  message?: string;
}

const negotiationService = {
  getOffersByBuyer: (buyerId: number) =>
    api.get<Offer[]>(`/api/negotiation/offers/buyer/${buyerId}`).then(res => res.data),

  getOffersBySupplier: (supplierId: number) =>
    api.get<Offer[]>(`/api/negotiation/offers/supplier/${supplierId}`).then(res => res.data),

  getAllOffers: () =>
    api.get<Offer[]>('/api/negotiation/offers').then(res => res.data),

  createOffer: (data: CreateOfferPayload) =>
    api.post<Offer>('/api/negotiation/offers', data).then(res => res.data),

  acceptOffer: (id: number) =>
    api.put(`/api/negotiation/offers/${id}/accept`),

  rejectOffer: (id: number) =>
    api.put(`/api/negotiation/offers/${id}/reject`),

  counterOffer: (id: number, newPrice: number) =>
    api.put(`/api/negotiation/offers/${id}/counter`, null, {
      params: { newPrice },
    }),
};

export default negotiationService;

File: negotiation-service/src/main/java/ma/solostock/negotiation/service/NegotiationService.java
package ma.solostock.negotiation.service;
import lombok.RequiredArgsConstructor;
import ma.solostock.negotiation.dto.*;
import ma.solostock.negotiation.entity.*;
import ma.solostock.negotiation.repository.*;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;
import java.util.List;

@Service @RequiredArgsConstructor
public class NegotiationService {
    private final OfferRepository offerRepository;
    private final ContractRepository contractRepository;

    public Offer createOffer(OfferRequest req) {
        Offer offer = Offer.builder()
                .productId(req.getProductId()).productName(req.getProductName())
                .buyerId(req.getBuyerId()).supplierId(req.getSupplierId())
                .proposedPrice(req.getProposedPrice()).originalPrice(req.getOriginalPrice())
                .quantity(req.getQuantity()).message(req.getMessage()).build();
        return offerRepository.save(offer);
    }

    public Offer acceptOffer(Long id) {
        Offer offer = offerRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Offre non trouvée"));
        offer.setStatus(OfferStatus.ACCEPTED);
        offer.setUpdatedAt(LocalDateTime.now());
        return offerRepository.save(offer);
    }

    public Offer rejectOffer(Long id) {
        Offer offer = offerRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Offre non trouvée"));
        offer.setStatus(OfferStatus.REJECTED);
        offer.setUpdatedAt(LocalDateTime.now());
        return offerRepository.save(offer);
    }

    public Offer counterOffer(Long id, Double newPrice) {
        Offer offer = offerRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Offre non trouvée"));
        offer.setProposedPrice(newPrice);
        offer.setStatus(OfferStatus.COUNTER_OFFERED);
        offer.setUpdatedAt(LocalDateTime.now());
        return offerRepository.save(offer);
    }

    public Contract signContract(ContractRequest req) {
        Contract contract = Contract.builder()
                .offerId(req.getOfferId()).buyerId(req.getBuyerId())
                .supplierId(req.getSupplierId()).finalPrice(req.getFinalPrice())
                .quantity(req.getQuantity()).terms(req.getTerms())
                .startDate(req.getStartDate()).endDate(req.getEndDate()).build();
        return contractRepository.save(contract);
    }

    public List<Offer> getAllOffers() { return offerRepository.findAll(); }
    public List<Offer> getOffersByBuyer(Long buyerId) { return offerRepository.findByBuyerId(buyerId); }
    public List<Offer> getOffersBySupplier(Long supplierId) { return offerRepository.findBySupplierId(supplierId); }
    public List<Contract> getAllContracts() { return contractRepository.findAll(); }
    public long countOffers() { return offerRepository.count(); }
    public long countAccepted() { return offerRepository.countByStatus(OfferStatus.ACCEPTED); }
}